app/models/subscription.rb

```diff
@@ -69,4 +71,41 @@ def create_alert_run
   def expired?
     expires_on < Time.zone.today
   end
+
+  private
+
+  def needs_reference?
+    new_record? && reference.blank?
+  end
+
+  def keyword_reference_part
+    search_criteria_to_h if @search_criteria_hash.nil?
+
+    @search_criteria_hash['keyword'].strip.split(/\s+/).join(' ') if @search_criteria_hash.key?('keyword')
+  end
+
+  def location_reference_part
+    search_criteria_to_h if @search_criteria_hash.nil?
+
+    has_location = ['location', 'radius'].all? { |key| @search_criteria_hash.key?(key) }
+
+    "within #{@search_criteria_hash['radius']} miles of #{@search_criteria_hash['location'].strip}" if has_location
+  end
+
+  def default_reference
```

Comment:
I think this would be better broken out into a service object (in `app/services`). Will keep the model code nice and tidy and allow it to be more easily testable.

Reply:
Done.